sycl/source/detail/device_filter.cpp

```diff
@@ -68,15 +66,24 @@ device_filter::device_filter(const std::string &FilterString) {
                              std::end(SyclDeviceTypeMap), findElement);
     // If no match is found, set device_type 'all',
     // which actually means 'any device_type' will be a match.
-    DeviceType = selectElement(Iter, SyclDeviceTypeMap, info::device_type::all);
+    if (Iter == SyclDeviceTypeMap.end())
+      DeviceType = info::device_type::all;
+    else {
+      DeviceType = Iter->second;
+      ColonPos = FilterString.find(":", Cursor);
+      if (ColonPos != std::string::npos)
+        Cursor = ColonPos + 1;
+      else
+        Cursor = Cursor + Iter->first.size();
+    }
   }
 
   // Handle the optional 3rd field of the filter, device number
   // Try to convert the remaining string to an integer.
   // If succeessful, the converted integer is the desired device num.
   if (Cursor < FilterString.size()) {
     try {
-      DeviceNum = stoi(FilterString.substr(ColonPos + 1));
+      DeviceNum = stoi(FilterString.substr(Cursor));
```

Comment:
Changes above (lambda inlining) are explained in the description. Could you please explain what this change is about? This looks like some additional change.

Reply:
This change was needed to handle the cases where SYCL_DEVICE_FILTER is set to only device numbers.
For example, SYCL_DEVICE_FILTER=0 does not have a colon (:) in the FilterString.

Reply:
Is it possible to add a test for this change?

Reply:
Done.